compiler/src/dotty/tools/dotc/util/Experimental.scala

```diff
@@ -0,0 +1,32 @@
+package dotty.tools.dotc
+package util
+
+import dotty.tools.dotc.ast.tpd
+import dotty.tools.dotc.ast.Trees._
+import dotty.tools.dotc.config.Feature
+import dotty.tools.dotc.core.Contexts._
+import dotty.tools.dotc.core.Symbols._
+import dotty.tools.dotc.core.Types._
+import dotty.tools.dotc.core.Flags._
+import dotty.tools.dotc.transform.SymUtils._
+
+object Experimental:
+  import tpd._
+
+  def checkExperimental(tree: Tree)(using Context): Unit =
+    if tree.symbol.isExperimental
+      && !tree.symbol.isConstructor // already reported on the class
+      && !tree.symbol.is(ModuleClass) // already reported on the module
+      && (tree.span.exists || tree.symbol != defn.ExperimentalAnnot) // already reported on inferred annotations
+    then
+      Feature.checkExperimentalFeature(tree.symbol.show, tree)
```

Comment:
We should avoid calling `show` if we're not going to actually show anything, I suggest having a `checkExperimentalSymbol` method that takes tree.symbol instead.